docs/painless/painless-guide/painless-datetime.asciidoc

```diff
@@ -334,3 +334,230 @@ if (zdt1.isAfter(zdt2)) {
+or `ctx['_source']`.
+
+*Examples*
+
+* Parse a numeric datetime from a sourced document to a complex datetime
++
+** Input:
++
+[source,JSON]
+----
+{
+    ...
+    "input_datetime": 434931327000
+    ...
+}
+----
++
+** Script:
++
+[source,Painless]
+----
+long inputDatetime = _source['input_datetime'];
+Instant instant = Instant.ofEpochMilli(inputDateTime);
+ZonedDateTime zdt = ZonedDateTime.ofInstant(instant, ZoneId.of('Z'));
+----
++
+* Parse a string datetime from a sourced document to a complex datetime
++
+** Input:
++
+[source,JSON]
+----
+{
+    ...
+    "input_datetime": "1983-10-13T22:15:30Z"
+    ...
+}
+----
++
+** Script:
++
+[source,Painless]
+----
+String datetime = ctx['_source']['input_datetime'];
+ZonedDateTime zdt = ZonedDateTime.parse(datetime);
+----
+Note the parse method uses ISO 8601 by default.
+
+===== Datetime Input From an Indexed Document
+
+Use an indexed document as a script input to access a complex datetime for a
+specific field within that document where the field is mapped as a
+{ref}/date.html[standard date] or a {ref}/date_nanos.html[nanosecond date].
+Numeric datetime fields mapped as {ref}/number.html[numeric] and string
+datetime fields mapped as {ref}/keyword.html[keyword] are accessible through an
+indexed document as well. Access to an indexed document within a script is
+dependent on the Painless context and is not always available. An indexed
+document is most commonly accessible through an input called `doc`.
+
+*Examples*
+
+* Format a complex datetime from an indexed document to a string datetime
++
+** Assumptions:
++
+*** The field `input_datetime` exists in all indexes as part of the query
+*** All indexed documents contain the field `input_datetime`
++
+** Mappings:
++
+[source,JSON]
+----
+{
+    "mappings": {
+        ...
+        "properties": {
+            ...
+            "input_datetime": {
+                "type": "date"
+            }
+            ...
+        }
+        ...
+    }
+}
+----
++
+** Script:
++
+[source,Painless]
+----
+def inputDatetime = doc['input_datetime'].value;
+String datetime = inputDateTime.format(DateTimeFormatter.ISO_INSTANT);
+----
+Note the use of a built-in DateTimeFormatter.
++
+* Find the difference between two complex datetimes from an indexed document
++
+** Assumptions:
++
+*** The fields `start_datetime` and `end_datetime` may *not* exist in all
+indexes as part of the query
+*** The fields `start_datetime` and `end_datetime` may *not* have values in all
+indexed documents
++
+** Mappings:
++
+[source,JSON]
+----
+{
+    "mappings": {
+        ...
+        "properties": {
+            ...
+            "start_datetime": {
+                "type": "date"
+            },
+            "end_datetime": {
+                "type": "date"
+            }
+            ...
+        }
+        ...
+    }
+}
+----
++
+** Script:
++
+[source,Painless]
+----
+// check for the existence of fields within the current document
```

Comment:
The contains key check is for existence of a mapping. This is relevant when doing a multi index search. If hte mappinng exists, doc will have the field, it just might not have any values.

Reply:
I understand. I'll try to reword this to make it clear what is happening.